### ISSUE TITLE
Trained Yolo Model doesnt reconize anything(even if given image was part of its training data)

### ISSUE THREAD

**Issue by Benz-Tracxpoint** [AlexeyAB/darknet#2038]
I have following AlexeyAB guide on training my custom data for YOLO V3.
I have created a huge custom dataset of 89 classes.
I have checked it and validate a few images and bounding boxes using the "Yolo_mark" tool -
all seems good - bounding boxes were correct and classes were as expected
I have trained based on Yolo V3 weights using the following command:

> ./darknet detector train cfg/my.data cfg/my.cfg yolov3.weights labels=cfg/my.names -gpus 0 -dont_show  -clear 1

afterwards I have taken the 1000 steps weights and trained the rest of my training data using it as base (as suggested for using multiple GPUs training):

> ./darknet detector train cfg/my.data cfg/my.cfg my_1000.weights labels=cfg/my.names -gpus 0,1,2,3 -dont_show  -clear 1

I have gotten as far as: my_62720.weights
in the few last thousands of training steps i kept getting the nan predictions such as:

>     52786: -nan, nan avg loss, 0.004000 rate, 19.859683 seconds, 13513216 images
>     Loaded: 0.000360 seconds
>     Region 82 Avg IOU: nan, Class: nan, Obj: nan, No Obj: nan, .5R: 0.000000, .75R: 0.000000,  count: 24
>     Region 82 Avg IOU: nan, Class: nan, Obj: nan, No Obj: nan, .5R: 0.000000, .75R: 0.000000,  count: 32
>     Region 82 Avg IOU: nan, Class: nan, Obj: nan, No Obj: nan, .5R: 0.000000, .75R: 0.000000,  count: 30
>     Region 82 Avg IOU: nan, Class: nan, Obj: nan, No Obj: nan, .5R: 0.000000, .75R: 0.000000,  count: 50

When trying to predict using the weights i have created I am getting a few weird issues
first: I have gotten this weird error message:

>  Error: in the file data/coco.names number of names 80 that isn't equal to classes=89 in the file cfg/my.cfg 

So I tried to manually override this weird issue by renaming the my.names to data/coco.names
afterwards I didnt get any detections and the prediction image was exactly the same as the original image(either for training images or for images that were never used in the training set)

here is the content of "my.data":

> classes= 89
> train  = /data/syntheticMasks/train.txt
> valid  = /data/syntheticMasks/test.txt
> names = /data/syntheticMasks/darknet/cfg/my.names
> backup = /data/syntheticMasks/backup/
> 

any ideas or suggestions on what did i do wrong?

**Comment by AlexeyAB** [comment]
@Benz-Tracxpoint Hi,

> ./darknet detector train cfg/my.data cfg/my.cfg yolov3.weights labels=cfg/my.names -gpus 0 -dont_show -clear 1

As desribed here you should use `darknet53.conv.74` pre-trained file instead of `yolov3.weights`: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

`darknet.exe detector train data/obj.data yolo-obj.cfg darknet53.conv.74`

----

> ./darknet detector train cfg/my.data cfg/my.cfg yolov3.weights labels=cfg/my.names -gpus 0 -dont_show -clear 1

What does it mean `labels=cfg/my.names` ?

----

> When trying to predict using the weights i have created I am getting a few weird issues
> first: I have gotten this weird error message:
> 
> Error: in the file data/coco.names number of names 80 that isn't equal to classes=89 in the file cfg/my.cfg

As described here: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> Note: After training use such command for detection: darknet.exe detector test data/obj.data yolo-obj.cfg yolo-obj_8000.weights

You should use test command like this:
`./darknet detector test cfg/my.data cfg/my.cfg my_10000.weights dog.jpg`

instead of 
`./darknet detector test cfg/coco.data cfg/my.cfg my_10000.weights dog.jpg`

----

* Do you use this repository https://github.com/AlexeyAB/darknet or this https://github.com/pjreddie/darknet ?

* What is the date of your Darknet code, when you clone it?

* What parameters do you use in the Makefile?

* And what GPU do you use?

* What versions of CUDA and cuDNN do you use? Does cuDNN version correspond to the CUDA version? You shouldn't use cuDNN 7.3 for CUDA 9.0 with CUDA 10.0

* Try to check mAP for `my_10000.weights` as described here: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

`./darknet detector map cfg/my.data cfg/my.cfg my_10000.weights`
and 
`./darknet detector map cfg/my.data cfg/my.cfg my_20000.weights`


----

**Comment by Benz-Tracxpoint** [comment]
Hi,
Thanks for the quick response!

> > ./darknet detector train cfg/my.data cfg/my.cfg yolov3.weights labels=cfg/my.names -gpus 0 -dont_show -clear 1

> As desribed here you should use `darknet53.conv.74` pre-trained file instead of `yolov3.weights`: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> `./darknet detector train data/obj.data yolo-obj.cfg darknet53.conv.74`

> > ./darknet detector train cfg/my.data cfg/my.cfg yolov3.weights labels=cfg/my.names -gpus 0 -dont_show -clear 1

> > What does it mean `labels=cfg/my.names` ?

For my next run i will use the following:
`./darknet detector train cfg/my.data cfg/my.cfg darknet53.conv.74  -gpus 0 -dont_show -clear 1`
Am I training with multiple GPUs in the correct way? waiting for 1000 steps weights and then call the command:
`./darknet detector train cfg/my.data cfg/my.cfg my_1000.weights -gpus 0,1,2,3 -dont_show -clear 1`
?

> > When trying to predict using the weights i have created I am getting a few weird issues
> > first: I have gotten this weird error message:
> > Error: in the file data/coco.names number of names 80 that isn't equal to classes=89 in the file cfg/my.cfg

> > As described here: https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects

> > Note: After training use such command for detection: darknet.exe detector test data/obj.data yolo-obj.cfg yolo-obj_8000.weights

> You should use test command like this:
> `./darknet detector test cfg/my.data cfg/my.cfg my_10000.weights dog.jpg`

Tried it with the weights files i have - still no detection, i will try it after running the training again with the advices weights file today.

> * Do you use this repository https://github.com/AlexeyAB/darknet or this https://github.com/pjreddie/darknet ?
> * What is the date of your Darknet code, when you clone it?

This is the last log i have checked out:
>	commit 21a4ec9390b61c0baa7ef72e72e59fa143daba4c
>	Author: AlexeyAB <alexeyab84@gmail.com>
>	Date:   Mon Nov 26 11:11:56 2018 +0300
>
>		Saving loss-chart for each 100 iterations automatically

> * What parameters do you use in the Makefile?
Not sure, the default ones.
> * And what GPU do you use?
I am training it on 4 x Tesla V100 GPUs

> * What versions of CUDA and cuDNN do you use? Does cuDNN version correspond to the CUDA version? You shouldn't use cuDNN 7.3 for CUDA 9.0 with CUDA 10.0
CUDNN 7.4.0, CUDA: V10.0.130

> * Try to check mAP for `my_10000.weights` as described here: https://github.com/AlexeyAB/darknet#when-should-i-stop-training

> `./darknet detector map cfg/my.data cfg/my.cfg my_10000.weights`
and
> `./darknet detector map cfg/my.data cfg/my.cfg my_20000.weights`

So if I have 89 classes I am looking for 89*2000 iterations in total(178,000) right? I will try using this command during the next training session

**Comment by AlexeyAB** [comment]
> Am I training with multiple GPUs in the correct way? waiting for 1000 steps weights and then call the command:
> ./darknet detector train cfg/my.data cfg/my.cfg my_1000.weights -gpus 0,1,2,3 -dont_show -clear 1
> ?

Yes.
Just use `-clear` instead of `-clear 1`.
And if you don't use yolov3.weights, then don't use `-clear` at all.

----

> This is the last log i have checked out:

Download the newest code from this repository. There was added fix for `Nan` bugs when you compiled with CUDNN_HALF=1: https://github.com/AlexeyAB/darknet/commits/master

----

> What parameters do you use in the Makefile?
> Not sure, the default ones.

By default in the Makefile all parameters are 0: https://github.com/AlexeyAB/darknet/blob/742bb7c7ce51a9faad08dc01fbf2e6908ad5ba4b/Makefile#L1-L7
So by default it will not use GPU at all, and training for ~100 000 iterations will take about 10 years )
So I think you set GPU=1 CUDNN=1 CUDNN_HALF=1.

----

**Update your code from this repository.** 
Set 
```
GPU=1
CUDNN=1
CUDNN_HALF=1
OPENCV=1

DEBUG=0
```
in the `Makefile`.
**compile**.

and train by using these commands:
`./darknet detector train cfg/my.data cfg/my.cfg darknet53.conv.74 -gpus 0 -dont_show`

`./darknet detector train cfg/my.data cfg/my.cfg backup/my_1000.weights -gpus 0,1,2,3 -dont_show`

**Note:** If during training you see `nan` values for avg (loss) field - then training goes wrong, but if `nan ` is in some other lines - then training goes well.

Then after training check mAP:
`./darknet detector map cfg/my.data cfg/my.cfg backup/my_10000.weights`
`./darknet detector map cfg/my.data cfg/my.cfg backup/my_100000.weights`

**Comment by Benz-Tracxpoint** [comment]
Hi,
I have trained a 1000 weights following your instructions(with the new version of the code), afterwards trying to train the rest of my training session based on this I am getting nan values as soon as the 3rd or 4th training step -
> 1008: inf, inf avg loss, 0.002000 rate, 6.043353 seconds, 129024 images
> Loaded: 0.000067 seconds
> Region 82 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 51
> Region 82 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 44
> Region 94 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 16
> Region 94 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 9
> Region 106 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 1
> Region 106 Avg IOU: -nan, Class: -nan, Obj: -nan, No Obj: 0.000000, .5R: -nan, .75R: -nan,  count: 0
> Region 82 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 58
> Region 82 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  cou
IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 15
Region 94 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 11
Region 106 Avg IOU: nan, Class: 0.000000, Obj: 0.000000, No Obj: 0.000000, .5R: 0.000000, .75R: 0.000000,  count: 1nt: 62
Region 94 Avg

**Comment by AlexeyAB** [comment]
Try to train first 5000 iterations with CUDNN_HALF=0
`./darknet detector train cfg/my.data cfg/my.cfg darknet53.conv.74 -gpus 0 -dont_show`

And then continue train with CUDNN_HALF=1
`./darknet detector train cfg/my.data cfg/my.cfg backup/my_5000.weights -gpus 0,1,2,3 -dont_show`

**Comment by Benz-Tracxpoint** [comment]
Hi - still the same issue occurrs.
I am starting to thing that i didnt make the cfg right for my 89 classes,
I am attaching it - I believe that i have following your directions but i might have missed something...

[my.txt](https://github.com/AlexeyAB/darknet/files/2667585/my.txt)
(changed the name to .txt for the upload)

**Comment by AlexeyAB** [comment]
@Benz-Tracxpoint Hi,

Your cfg file is ok.

* After how many iterations do you get `nan` for avg loss?

* Is there created files `bad.list` or `bad_label.list` near with darknet executable files?

* Can you train successfully your model with `CUDNN_HALF=0` and  `-gpus 0,1,2,3`?

* Or can you train successfully your model with `CUDNN_HALF=1` and `-gpus 0`?

**Comment by Benz-Tracxpoint** [comment]
With CUDNN_HALF=1 I barely get to 500 iterations
without it It feels like about 5000 6000 regardless of the gpu amount
ant no the files werent created

**Comment by AlexeyAB** [comment]
> With CUDNN_HALF=1 I barely get to 500 iterations

Download the newset version of Darknet from this repository, and re-compile it with `CUDNN_HALF=1`.
In the new version Fp16 is disabled for the first 2000 iterations.

----

> without it It feels like about 5000 6000 regardless of the gpu amount

Try to set `CUDNN_HALF=0`, recompile and run 
`./darknet detector train cfg/my.data cfg/my.cfg darknet53.conv.74 -dont_show`

Train about 10 000 iterations, if `nan` value appears, then check your cuDNN version and your dataset.
